File: apps/backend-node/src/features/admin/admin.controller.ts
import {
  Body,
  Controller,
  Get,
  Param,
  ParseUUIDPipe,
  Patch,
  Post,
} from '@nestjs/common';

import type { AuthUser } from '../../shared/auth/auth-user.model';
import { RequireUser } from '../../shared/auth/require-user.decorator';
import type { HostBookingDto } from '../hosts/dto/host-booking.dto';
import type { HostDto } from '../hosts/dto/host.dto';
import type { PropertyStatusChange } from '../properties/dto/property-status-change';
import { AdminService } from './admin.service';
import type {
  AdminDashboardStatsDto,
  AdminHostDto,
  AdminPaymentDto,
  AdminPropertyDto,
  AuditLogDto,
  HousingRequestDto,
  UpdateHousingRequestStatusDto,
} from './dto';

@Controller('admin')
export class AdminController {
  constructor(private readonly adminService: AdminService) {}

  @Get('stats')
  getAdminStats(
    @RequireUser() user: AuthUser,
  ): Promise<AdminDashboardStatsDto> {
    return this.adminService.getDashboardStats(user);
  }

  @Get('properties')
  listAdminProperties(
    @RequireUser() user: AuthUser,
  ): Promise<AdminPropertyDto[]> {
    return this.adminService.listProperties(user);
  }

  @Post('properties/:propertyId/publish')
  publishProperty(
    @RequireUser() user: AuthUser,
    @Param('propertyId', ParseUUIDPipe) propertyId: string,
  ): Promise<PropertyStatusChange> {
    return this.adminService.publishProperty(user, propertyId);
  }

  @Post('properties/:propertyId/reject')
  rejectPropertyReview(
    @RequireUser() user: AuthUser,
    @Param('propertyId', ParseUUIDPipe) propertyId: string,
  ): Promise<PropertyStatusChange> {
    return this.adminService.rejectPropertyReview(user, propertyId);
  }

  @Get('hosts')
  listAdminHosts(@RequireUser() user: AuthUser): Promise<AdminHostDto[]> {
    return this.adminService.listHosts(user);
  }

  @Post('hosts/:hostId/approve')
  approveHost(
    @RequireUser() user: AuthUser,
    @Param('hostId', ParseUUIDPipe) hostId: string,
  ): Promise<HostDto> {
    return this.adminService.approveHost(user, hostId);
  }

  @Get('bookings')
  listAdminBookings(
    @RequireUser() user: AuthUser,
  ): Promise<HostBookingDto[]> {
    return this.adminService.listBookings(user);
  }

  @Get('payments')
  listAdminPayments(
    @RequireUser() user: AuthUser,
  ): Promise<AdminPaymentDto[]> {
    return this.adminService.listPayments(user);
  }

  @Get('housing-requests')
  listHousingRequests(
    @RequireUser() user: AuthUser,
  ): Promise<HousingRequestDto[]> {
    return this.adminService.listHousingRequests(user);
  }

  @Patch('housing-requests/:requestId')
  updateHousingRequestStatus(
    @RequireUser() user: AuthUser,
    @Param('requestId', ParseUUIDPipe) requestId: string,
    @Body() request: UpdateHousingRequestStatusDto,
  ): Promise<HousingRequestDto> {
    return this.adminService.updateHousingRequestStatus(user, requestId, request);
  }

  @Get('audit-logs')
  listAuditLogs(@RequireUser() user: AuthUser): Promise<AuditLogDto[]> {
    return this.adminService.listAuditLogs(user);
  }
}

File: apps/backend-java/src/main/java/com/housingplatform/admin/AdminController.java
package com.housingplatform.admin;

import com.housingplatform.auth.AuthSupport;
import com.housingplatform.admin.dto.AdminDashboardStats;
import com.housingplatform.admin.dto.AdminHost;
import com.housingplatform.admin.dto.AdminPayment;
import com.housingplatform.admin.dto.AdminProperty;
import com.housingplatform.admin.dto.AuditLogDto;
import com.housingplatform.admin.dto.HousingRequestDto;
import com.housingplatform.admin.dto.UpdateHousingRequestStatusRequest;
import com.housingplatform.hosts.dto.HostBooking;
import com.housingplatform.hosts.dto.HostDto;
import com.housingplatform.properties.dto.PropertyStatusChange;
import jakarta.validation.Valid;
import java.util.List;
import java.util.UUID;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("${housing-platform.api-prefix}/admin")
public class AdminController {

  private final AdminService adminService;

  public AdminController(AdminService adminService) {
    this.adminService = adminService;
  }

  @GetMapping("/stats")
  public AdminDashboardStats getAdminStats() {
    return adminService.getDashboardStats(AuthSupport.requireCurrentUser());
  }

  @GetMapping("/properties")
  public List<AdminProperty> listAdminProperties() {
    return adminService.listProperties(AuthSupport.requireCurrentUser());
  }

  @PostMapping("/properties/{propertyId}/publish")
  public PropertyStatusChange publishProperty(@PathVariable UUID propertyId) {
    return adminService.publishProperty(AuthSupport.requireCurrentUser(), propertyId);
  }

  @PostMapping("/properties/{propertyId}/reject")
  public PropertyStatusChange rejectPropertyReview(@PathVariable UUID propertyId) {
    return adminService.rejectPropertyReview(AuthSupport.requireCurrentUser(), propertyId);
  }

  @GetMapping("/hosts")
  public List<AdminHost> listAdminHosts() {
    return adminService.listHosts(AuthSupport.requireCurrentUser());
  }

  @PostMapping("/hosts/{hostId}/approve")
  public HostDto approveHost(@PathVariable UUID hostId) {
    return adminService.approveHost(AuthSupport.requireCurrentUser(), hostId);
  }

  @GetMapping("/bookings")
  public List<HostBooking> listAdminBookings() {
    return adminService.listBookings(AuthSupport.requireCurrentUser());
  }

  @GetMapping("/payments")
  public List<AdminPayment> listAdminPayments() {
    return adminService.listPayments(AuthSupport.requireCurrentUser());
  }

  @GetMapping("/housing-requests")
  public List<HousingRequestDto> listHousingRequests() {
    return adminService.listHousingRequests(AuthSupport.requireCurrentUser());
  }

  @PatchMapping("/housing-requests/{requestId}")
  public HousingRequestDto updateHousingRequestStatus(
      @PathVariable UUID requestId, @Valid @RequestBody UpdateHousingRequestStatusRequest request) {
    return adminService.updateHousingRequestStatus(
        AuthSupport.requireCurrentUser(), requestId, request);
  }

  @GetMapping("/audit-logs")
  public List<AuditLogDto> listAuditLogs() {
    return adminService.listAuditLogs(AuthSupport.requireCurrentUser());
  }
}
